### PR TITLE
HRBE:56 - Created the views and serializers for login api

### PR DESCRIPTION
**Pull request by deepak-jv**

### CODE REVIEW

File: accounts/views.py [hunk 1]
```diff
@@ -1,3 +1,14 @@
 from django.shortcuts import render
+from requests import Response
+from rest_framework import generics, status
 
-# Create your views here.
+from accounts.serializers import LoginSerializer
+
+
+class LoginAPIView(generics.GenericAPIView):
```
Comment (Abishek05) [inline accounts/views.py:8]: use ModelViewset please

File: accounts/serializers.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        if filtered_user_by_email.exists() and filtered_user_by_email[0].auth_provider != 'email':
+            raise AuthenticationFailed(
+                detail='Please continue your login using ' + filtered_user_by_email[0].auth_provider)
+
+        if not user:
+            raise AuthenticationFailed('Invalid credentials, try again')
+        if not user.is_active:
+            raise AuthenticationFailed('Account disabled, contact admin')
+        if not user.is_verified:
```
Comment (Abishek05) [inline accounts/serializers.py:42]: This should be `is_email_verified` as per the User model

File: accounts/serializers.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    def validate(self, attrs):
+        email = attrs.get('email', '')
+        password = attrs.get('password', '')
+        filtered_user_by_email = User.objects.filter(email=email)
+        user = auth.authenticate(email=email, password=password)
+
+        if filtered_user_by_email.exists() and filtered_user_by_email[0].auth_provider != 'email':
+            raise AuthenticationFailed(
+                detail='Please continue your login using ' + filtered_user_by_email[0].auth_provider)
```
Comment (Abishek05) [inline accounts/serializers.py:36]: use f string if possible

File: accounts/serializers.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+from accounts.models import User
+from rest_framework.exceptions import AuthenticationFailed
+
+
+class LoginSerializer(serializers.ModelSerializer):
+    email = serializers.EmailField(max_length=255, min_length=3)
+    password = serializers.CharField(
+        max_length=68, min_length=6, write_only=True)
+    username = serializers.CharField(
+        max_length=255, min_length=3, read_only=True)
```
[the remark this answers is not in the document]
Comment (deepak-jv) [reply accounts/serializers.py]: done

File: accounts/views.py (hunk 1 above)
Comment (deepak-jv) [reply accounts/views.py]: done

File: accounts/serializers.py (hunk 2 above)
Comment (deepak-jv) [reply accounts/serializers.py]: done

File: accounts/serializers.py (hunk 3 above)
Comment (deepak-jv) [reply accounts/serializers.py]: done

File: accounts/views.py [hunk 5]
```diff
@@ -1,3 +1,14 @@
...
+    serializer_class = LoginSerializer
+
+    def post(self, request):
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
[the remark this answers is not in the document]
Comment (deepak-jv) [reply accounts/views.py]: done

File: accounts/serializers.py [hunk 6]
```diff
@@ -1,7 +1,50 @@
...
+                raise serializers.ValidationError("User account is disabled.")
+            user = authenticate(
+                request=self.context.get("request"), email=email, password=password
+            )
+            if user:
+                if not user.is_email_verified:
+                    raise serializers.ValidationError("Email is not verified.")
+
+                if not user.is_active:
+                    raise serializers.ValidationError("User account is disabled.")
```
Comment (Abishek05) [inline accounts/serializers.py:35]: User account is disabled is used twice

File: accounts/views.py [hunk 7]
```diff
@@ -1,9 +1,37 @@
...
+from .serializers import SkillSerializer, LoginSerializer
 from .permissions import IsAdminOrReadOnlyCustom
 
-# Create your views here.
+
+class LoginViewSet(viewsets.ModelViewSet):
+    queryset = User.objects.all()
+    serializer_class = LoginSerializer
+    http_method_names = [
+        "post",
```
Comment (Abishek05) [inline accounts/views.py:15]: inclue "head" and "options"

File: accounts/views.py [hunk 8]
```diff
@@ -1,9 +1,37 @@
...
+from rest_framework import viewsets
 from .models import Skill
-from .serializers import SkillSerializer
+from .serializers import SkillSerializer, LoginSerializer
 from .permissions import IsAdminOrReadOnlyCustom
 
-# Create your views here.
+
+class LoginViewSet(viewsets.ModelViewSet):
+    queryset = User.objects.all()
```
Comment (Abishek05) [inline accounts/views.py:12]: Queryset is not required as we are disabling "GET" method

File: accounts/serializers.py [hunk 9]
```diff
@@ -1,7 +1,50 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth import authenticate
 from rest_framework import serializers
 from .models import Skill
 
+UserModel = get_user_model()
+
+
+class LoginSerializer(serializers.Serializer):
```
Comment (Abishek05) [inline accounts/serializers.py:9]: Use ModelSerializer and ```extra_kwargs = {
            'password': {
                'write_only': True,
                'style': {'input_type': 'password'}
            },
    
        }```
to make sure passsword is hidden in browsable API

File: accounts/views.py [hunk 10]
```diff
@@ -1,9 +1,37 @@
...
+        access_token = str(refresh.access_token)
+        refresh_token = str(refresh)
+
+        return Response(
+            {
+                "access_token": access_token,
+                "refresh_token": refresh_token,
+            },
+            status=status.HTTP_200_OK,
+        )
```
Comment (Abishek05) [inline accounts/views.py:34]: This logic , please put it in serilaizer overriding validate() method

**Review by Abishek05 — CHANGES_REQUESTED** [review]
Seems like your IDE has automatically replaced all single quotes with double. Make sure to disable such extensions else your PR will have lot of changes and difficult to review

File: mysite/urls.py [hunk 11]
```diff
@@ -16,29 +16,38 @@
...
 from rest_framework import permissions
 from drf_yasg.views import get_schema_view
 from drf_yasg import openapi
 
-
 router = DefaultRouter()
-router.register(r"skill-dropdown", SkillViewSet, basename="skill")
+router.register(r"skill-dropdown", SkillViewSet, basename="skill")  # noqa: E501
+router.register("login", LoginViewSet, basename="login")  # noqa : E501
+router.register(r"skill-dropdown", SkillViewSet, basename="skill")  # noqa: E501
```
Comment (Abishek05) [inline mysite/urls.py:28]: remove `r`, it's not required

File: accounts/views.py (hunk 10 above)
Comment (deepak-jv) [reply accounts/views.py]: done

File: accounts/serializers.py (hunk 9 above)
Comment (deepak-jv) [reply accounts/serializers.py]: done

File: accounts/views.py (hunk 8 above)
Comment (deepak-jv) [reply accounts/views.py]: done

File: accounts/serializers.py (hunk 6 above)
Comment (deepak-jv) [reply accounts/serializers.py]: done

File: accounts/views.py (hunk 7 above)
Comment (deepak-jv) [reply accounts/views.py]: done

File: mysite/urls.py (hunk 11 above)
Comment (deepak-jv) [reply mysite/urls.py]: done

**Review by Abishek05 — APPROVED** [review]
LGTM

File: accounts/models.py [hunk 12]
```diff
@@ -91,6 +91,9 @@ class User(AbstractUser):
     USERNAME_FIELD = "email"
     REQUIRED_FIELDS = ["username", "password"]
 
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["password", "username"]
+
```
Comment (Abishek05) [inline accounts/models.py:96]: @deepak-jv Why was this added twice?

Comment (Abishek05) [reply accounts/models.py:96]: We don't require to set these

Comment (deepak-jv) [reply accounts/models.py:96]: My PR and Vasant's PR both gets merged on the same day and I got the merge conflicts but the you we have to show the demo on Friday that's why I might miss this to remove.

**Pull request merged**